src/inet/UDPEndPoint.cpp

```diff
@@ -543,6 +543,31 @@ INET_ERROR UDPEndPoint::SendMsg(const IPPacketInfo * pktInfo, System::PacketBuff
 
 #if CHIP_SYSTEM_CONFIG_USE_LWIP
 
+    pbuf * msgCopy = NULL;
+
+    if (msg.GetRef() > 1)
+    {
+        // when retaining a buffer, the caller expects the msg to be
+        // unmodified.  LwIP stack will normally prepend the packet
+        // headers as the packet traverses the UDP/IP/netif layers,
+        // which normally modifies the packet.  We prepend a small
+        // pbuf to the beginning of the pbuf chain, s.t. all headers
+        // are added to the temporary space, just large enough to hold
+        // the transport headers.
+        msgCopy = pbuf_alloc(PBUF_TRANSPORT, 0, PBUF_RAM);
+
+        if (msgCopy == NULL)
+        {
+            return INET_ERROR_NO_MEMORY;
+        }
+
+        pbuf_chain(msgCopy, System::LwIPPacketBufferView::UnsafeGetLwIPpbuf(msg));
+    }
+    else
+    {
+        msgCopy = System::LwIPPacketBufferView::UnsafeGetLwIPpbuf(msg);
+    }
+
```

Comment:
Overall this whole approach is very complex to follow and error prone, and tries to deal with *internal knowledge* of the IP stack (lwip here)

CRMP should make no attempt to keep the same buffer as that being sent in the retransmit table, or use reference counting, as a first implementation. The simpler and much less complex approach (which doesn't require special processing logic) is to:

1. Make CRMP keep the original copy for retransmit table, rather than increase refcount. Refcount increase are a total optimization with lots of side-effects in how you analyze downstream code.
2. Here, always make a copy, rather than chain. Make a literal, full copy of the chain.

You could do the opposite as well.

This would:

- Reduce special casing
- Create a copy that lives only for a very short period of time, only until the netif is reached, which is actually a synchronous flow for UDP
- Avoid having start tracking the exact chaining/ref count
- Avoid making assumption about the TCP/IP stack below (that its buffer enable and/or require particular reference count semantics and have underlying chaining considerations).

The cost of the copy is minuscule compared to many of the other costs, and *a copy needs to be made in some situations anyway*.

Reply:
I'm also concerned about exposing the reference count. I'll look deeper, but off the top of my head, what if we provide something like `PacketBufferHandle GetUniqueOwnership(PacketBufferHandle && aPacket)`, which returns the handle itself if ref==1 and a copy if ref>1? Then `UDPEndPoint::SendMsg()` can do this just before calling LwIP, and we don't have to know or expose any internals.

Reply:
@tcarmelveilleux Currently CRMP logic is implemented within the messaging layer and use the transport layer to uniquely encrypt and send the message. If we keep the original buffer in the re-transmit table, in the following re-transmission,  the original buffer will go through the encryption process again, then it will not be treated as the same packet by the destination.

I have piloted clone and send method In https://github.com/project-chip/connectedhomeip/pull/4927, I found the transport layer does not send the encrypted packet as is, it will re-attach the UDP header at least for each send, even I clone the whole encrypted packet, it will fail in the following re-transmission.

If we want to use copy and send method, the best place is do the copy is within UDPEndPoint::SendMsg(), but we need to update the current PacketBufferHandle CloneData() to support chained buffers.

    /**
     * Creates a copy of the data in this packet.
     *
     * Does NOT support chained buffers.
     *
     *  @param[in]  aAdditionalSize Size of additional application data space after the initial contents.
     *  @param[in]  aReservedSize   Number of octets to reserve for protocol headers.
     *
     * @returns empty handle on allocation failure.
     */
    PacketBufferHandle CloneData(uint16_t aAdditionalSize = 0, uint16_t aReservedSize = PacketBuffer::kDefaultHeaderReserve); 
    
Currently, the PacketBuffer is the wrapper of the head of pbuf, the Lwip does not provide the function to clone the pbuf chain. Since we don't own pbuf.c, implement the chained buffer copy within PacketBuffer will be very tricky.

Reply:
If we could assume we never use chained pbuf for CHIP message with Lwip stack(Currently YES, but not sure in future), then we can copy the  PacketBufferHandle at UDPEndPoint::SendMsg() and give the copy to Lwip stack to send. 

Confirmed the openweave also get updated to use chained pbuf method and verified in production.

Reply:
For a quick fix, we can make CloneData error out if a chain is passed in and have a GH issue to have chain support.

Reply:
Do we actually want `CloneData()` to handle chains, or a new `Clone()` that also duplicates ‘reserved’ (non Data) parts? I fear the former would end up biting someone.

Reply:
I registered https://github.com/project-chip/connectedhomeip/issues/5348 to follow up this, yes, clone a PacketBufferHandle with a chained pbuf will be super tricky from my point of view if we do not own pbuf.c, we need to make a copy of the head as PacketBuffer and rest of element as plain pbuf.

For now, I assume CHIP messages are never chained, CloneData will return failure if it find the buffer is chained.